File: CSharp/Store.cs
using System;

namespace ReduxKata
{
    class Store {
        private int state;
        private Func<int, Action, int> reducer;
        private Func<int, int> listener;

        public int getState() {
            return this.state;
        }
        public static Store createStore(Func<int, Action, int> reducer) {
            return new Store(reducer);
        }
        public int dispatch(Action action) {     
            this.state = this.reducer(this.state, action);
            listener(this.state);
            return this.state;
        }
        public void subscribe(Func <int, int> listener) {
            this.listener = listener;
        }
        
        Store(Func<int, Action, int> reducer) {
            this.reducer = reducer;
            this.state = 0;
        }
    }
}

File: Java/src/com/company/Store.java
package com.company;

import java.util.List;
import java.util.Vector;
import java.util.function.BiFunction;
import java.util.function.Function;

public class Store{
    private int state;
    private BiFunction<Integer, Action, Integer> reducer;
    private Vector<Listener> listeners;

    private Store(BiFunction<Integer, Action, Integer> reducer) {
        this.reducer = reducer;
        this.listeners = new Vector<Listener>();
        this.state = 0;
    }

    static Store createStore(BiFunction<Integer, Action, Integer> reducer) {
        return new Store(reducer);
    }

    public void subscribe(Function<Integer, Integer> listener) {

        this.listeners.add(new Listener(listener));
    }

    public Integer dispatch(Action action) {
        this.state = this.reducer.apply(this.state, action);
        for(Listener listener: listeners) {
            listener.apply(this.state);
        }
        return this.state;
    }

    private static class Listener {
        private Function<Integer, Integer> listener;

        public Listener(Function<Integer, Integer> listener) {
            this.listener = listener;
        }

        public void apply(Integer state) {
            this.listener.apply(state);
        }
    }
}

